### PR TITLE
Add explainer to captureTimestamp and senderCaptureTimeOffset in RTCR…

### PR DESCRIPTION
**Pull request by minyuel**

### CODE REVIEW

**Comment by minyuel** [comment]
Hi @henbos , please take a look!

File: explainer.md [hunk 1]
```diff
@@ -1,3 +1,118 @@
...
+
+The solution proposed in this document is based on a new RTP header extension, namely absolute capture time, which contains two data fields:
+ * an absolute capture timestamp, which is stamped by the original capturer, and is supposed to be received untouched by the end receivers.
+ * an estimated clock offset with respect to the capturer's clock, which is supposed to be read and updated by every intermediate sender.
+
+With the absolute capture timestamps, end receivers can accurately measure the audio-video synchronization performance. With the `estimated clock offset`, which gets updated at each intermediate hop, end receivers can estimate their respective clock offset with respect to the capturer's clock, and then together with the absolute capture timestamp, measure the end-to-end delay. 
+
+The absolute capture time RTP header extension has been adopted by WebRTC, see the specification [here](https://github.com/webrtc/webrtc-org/blob/gh-pages/experiments/rtp-hdrext/abs-capture-time/index.md).
```
Comment (henbos) [inline explainer.md]: "adopted by WebRTC"? If this is supposed to be referring to a specific implementation, perhaps reference "webrtc.org"?

However I think can just say "The RTP header extension is defined here:"

File: explainer.md [hunk 2]
```diff
@@ -1,3 +1,118 @@
...
+
+To surface the information to web applications, two data fields are proposed to be added to RTCRtpContributingSource, and has been adopted by WebRTC 1.0 extensions, see the specification [here](https://w3c.github.io/webrtc-extensions/#rtcrtpcontributingsource-dictionary).
```
Comment (henbos) [inline explainer.md]: I'd say "The WebRTC Extensions spec". This spec contains post-1.0 features.

However this explainer is on the spec in question's repo, so I would just say "This specification exposes (...the things it exposes...)"

File: explainer.md [hunk 3]
```diff
@@ -1,3 +1,118 @@
...
+We propose to add captureTimestamp, type of DOMHighResTimeStamp, to the RTCRtpContributingSource dictionary. This surfaces the absolute capture timestamp in the absolute capture time RTP header extension, when it is present or can be extrapolated from previously received data, for the last rendered audio or video frame. It can be used for measuring audio video synchronization performance as illustrated in the following exemplar code:
+
+    [receiverAudio, receiverVideo] = peerconnection.getReceivers();
+
+    latestCaptureTimestampsAudio = receiverAudio.getSynchronizationSources()[0].captureTimeStamp;
+    latestCaptureTimestampsVideo = receiverVideo.getSynchronizationSources()[0].captureTimeStamp;
+
+    avSynchronization = latestCaptureTimestampsAudio - latestCaptureTimestampsVideo;
+
+This new API has been accepted in WebRTC 1.0 extensions. See the specification [here](https://w3c.github.io/webrtc-extensions/#dom-rtcrtpcontributingsource-capturetimestamp).
```
Comment (henbos) [inline explainer.md]: The explainer of the spec that has accepted the feature does not need to say that it has been accepted by itself. Just say that "captureTimestamp is defined here" if you think a link is needed.

File: explainer.md [hunk 4]
```diff
@@ -1,3 +1,118 @@
...
+
+#### [API 2]: senderCaptureTimeOffset
+
+We also propose to add senderCaptureTimeOffset, type of DOMHighResTimeStamp, to the RTCRtpContributingSource dictionary. In this context, the sender refers to the system that directly sends RTP and RTCP packets to the receiver, and thus the sender-receiver path only represents the "last hop" in a system that involves intermediate stream regenerators.
```
Comment (henbos) [inline explainer.md]: "This specification adds..."

File: explainer.md [hunk 5]
```diff
@@ -1,3 +1,118 @@
 ## Explainer
```
Comment (henbos) [inline explainer.md]: Detailed and thorough document, very nice!

Can you just add a paragraph introducing the document as a whole. The spec is doing multiple different things, but this explainer makes it sound like it is all about capture timestamp.

I don't mean to write as thoroughly about the other stuff, I mean just an introduction. Basically what the [introduction section](https://w3c.github.io/webrtc-extensions/#introduction) is already saying.

**Comment by minyuel** [comment]
comments addressed, PTAL again.

File: explainer.md [hunk 6]
```diff
@@ -1,3 +1,122 @@
...
+chxg@google.com
+
+Participate
+https://www.chromestatus.com/feature/5728533701722112
+
+#### Abstract
+
+This section explains the two new data fields in RTCRtpContributingSource, namely captureTimestamp and senderCaptureTimeOffset, see [here](https://w3c.github.io/webrtc-extensions/#rtcrtpcontributingsource-dictionary). They are introduced for audio-video synchronization and end-to-end delay measurements.
+
+The solution to audio-video synchronization and end-to-end delay measurements described here is particularly desired by systems, where an intermediate steam regenerator that terminates the streams originating from senders, an audio mixer as an example, is involved.
```
Comment (alvestrand) [inline explainer.md]: Nit: no comma after "systems".
More readable: "systems where an intermediate stream regenerator that terminates the streams originating from senders is involved. One example of this is an audio mixer."

File: explainer.md [hunk 7]
```diff
@@ -1,3 +1,122 @@
...
+
+#### Introduction
+
+In simple real-time communication systems, which involves only a sender and a receiver, the RTCP scheme [RFC3550] allows estimation of round-trip time and thus end-to-end delay.
+However, in a more sophisticated system, where an intermediate stream regenerator is involved, the estimation of end-to-end delay becomes more difficult. For example, if audio mixing is performed on the server side, the server terminates inbound media streams, processes the media data, and then generates new outbound media streams with fresh synchronization information.
+
+The solution proposed in this document is based on a new RTP header extension, namely absolute capture time, which contains two data fields:
```
Comment (alvestrand) [inline explainer.md]: Delete "namely".

File: explainer.md [hunk 8]
```diff
@@ -1,3 +1,122 @@
...
+ * an absolute capture timestamp, which is stamped by the original capturer, and is supposed to be received untouched by the end receivers.
+ * an estimated clock offset with respect to the capturer's clock, which is supposed to be read and updated by every intermediate sender.
+
+With the absolute capture timestamps, end receivers can accurately measure the audio-video synchronization performance. With the `estimated clock offset`, which gets updated at each intermediate hop, end receivers can estimate their respective clock offset with respect to the capturer's clock, and then together with the absolute capture timestamp, measure the end-to-end delay.
```
Comment (alvestrand) [inline explainer.md]: Suggest replacing "the audio-video synchronization performance" with "how synchronized the audio and video tracks are". Target audience is not A/V experts.

Delete "respective".

Insert a comma after "and then" ("together with the absolute capture timestamp" is a subsidiary clause)

File: explainer.md [hunk 9]
```diff
@@ -1,3 +1,122 @@
...
+ * Facilitate calculation of audio video synchronization performance in real-time communication systems.
+ * Facilitate calculation of end-to-end delay in real-time communication systems, particularly those that involve intermediate stream regenerators.
+
+##### Non-goals
+
+This proposal does not aim for improving the accuracy of end-to-end delay measurement in real-time communication systems that can already measure it based on RTCP based schemes.
+
+#### [API 1]: captureTimestamp
+
+This specification adds captureTimestamp, type of DOMHighResTimeStamp, to the RTCRtpContributingSource dictionary. This surfaces the absolute capture timestamp in the absolute capture time RTP header extension, when it is present or can be extrapolated from previously received data, for the last rendered audio or video frame. It can be used for measuring audio video synchronization performance as illustrated in the following exemplar code:
```
Comment (alvestrand) [inline explainer.md]: exemplar -> example

File: explainer.md [hunk 10]
```diff
@@ -1,3 +1,122 @@
...
+
+    [receiverAudio, receiverVideo] = peerconnection.getReceivers();
+
+    latestCaptureTimestampsAudio = receiverAudio.getSynchronizationSources()[0].captureTimeStamp;
+    latestCaptureTimestampsVideo = receiverVideo.getSynchronizationSources()[0].captureTimeStamp;
+
+    avSynchronization = latestCaptureTimestampsAudio - latestCaptureTimestampsVideo;
```
Comment (alvestrand) [inline explainer.md]: suggest rename avSynchronization -> synchronizationError

File: explainer.md [hunk 11]
```diff
@@ -1,3 +1,122 @@
...
+
+#### [API 2]: senderCaptureTimeOffset
+
+This specification also adds senderCaptureTimeOffset, type of DOMHighResTimeStamp, to the RTCRtpContributingSource dictionary. In this context, the sender refers to the system that directly sends RTP and RTCP packets to the receiver, and thus the sender-receiver path only represents the "last hop" in a system that involves intermediate stream regenerators.
+
+An exemplar code to use captureTimestamp and senderCaptureTimeOffset to calculate end-to-end delay follows:
```
Comment (alvestrand) [inline explainer.md]: exemplar -> example
Remove "follows".

File: explainer.md [hunk 12]
```diff
@@ -1,3 +1,122 @@
...
+
+    receiver = peerconnection.getReceivers()[0];
+    csrc = receiver.getSynchronizationSources()[0];
+    latestCaptureTimestamps = csrc.captureTimeStamp;
```
Comment (alvestrand) [inline explainer.md]: latestCaptureTimestamps -> latestCaptureTimestamp (there is only one)

File: explainer.md [hunk 13]
```diff
@@ -1,3 +1,122 @@
...
+    endToEndDelay = receiverTimestamp - receiverCaptureTimestamp.
+
+#### Detailed design discussion
+
+##### [Tricky design choice 1]: multiple ways to surface end-to-end delay
+
+The proposed solution is, as [API 2], to surface the sender-capture time offset as the raw data existing the absolute capture time RTP header extension. Then the calculation of end-to-end delay requires other data, as depicted in the code example in [API 2]. Alternatively, we can surface another derived quality, even just the end-to-end delay. However, for the sake of clarity and testability in the specification, it would be best to report the raw data.
+
+#### Considered alternatives
+
```
Comment (alvestrand) [inline explainer.md]: You should probably start off this section by saying something like "We believe that given the choice of signaling protocol, the API suggested here is obvious. This section will sketch some of the other proposals for signaling protocols that have been considered and rejected over the development of this proposal."

File: explainer.md [hunk 14]
```diff
@@ -1,3 +1,122 @@
...
+However, without the original capture timestamp, this method may fail if the intermediate server regenerates the stream and applies RTP timestamping that does not strictly follow the capturer's clock rate. A discussion of this scheme can be found [here](https://github.com/w3c/webrtc-stats/issues/537).
+
+##### [Alternative 2] Audio Timing Header Extension
+
+In WebRTC, [video-timing](https://github.com/webrtc/webrtc-org/blob/gh-pages/experiments/rtp-hdrext/video-timing/index.md) has been proposed as an experimental RTP header extension. We reject the idea of reusing it for audio or adding an audio version of it, since it has a duration-based design that requires its header extension to be sent with every frame, hence consuming significantly more bandwidth than our timestamp-based design.
+
+##### [Alternative 3] [RFC5484]: SMPTE Time-Code
+
+We reject the proposal in [RFC5484] to use SMPTE time-codes. It seems needlessly complex for our use cases. It would also only help us solve the desynchronization metric problem, and not provide us with a solution for a one-way delay metric.
```
Comment (alvestrand) [inline explainer.md]: Should probably say "We rejected this proposal" rather than "We reject".

File: explainer.md [hunk 15]
```diff
@@ -1,3 +1,122 @@
...
+##### [Alternative 4] [RFC6051]: Rapid Synchronisation of RTP Flows
+
+We reject the proposals in [RFC6051] since they cannot overcome the problem of synchronizing beyond mixers. The extended version of abs-capture-time does however borrow design elements from the RFC’s In-Band Delivery of Synchronisation Metadata section.
+
+#### Stakeholder Feedback / Opposition
+
+The proposal has been presented to W3C WG and browser implementers. While there is no signal from other implementers to implement this yet, the proposal has been reviewed without concerns raised.
+
+Real applications, Google Hangouts and Meet, for example, have been asking for reliable audio video synchronization and end-to-end delay measurements for several years, which we can back up with [discussions](https://github.com/w3c/webrtc-stats/issues/158) from 2017.
+
```
Comment (alvestrand) [inline explainer.md]: Are they asking for a privacy and security considerations section?

File: explainer.md (hunk 12 above)
Comment (minyuel) [reply explainer.md]: Thanks!

File: explainer.md (hunk 15 above)
Comment (minyuel) [reply explainer.md]: not as I found in this template. https://w3ctag.github.io/explainers

**Comment by minyuel** [comment]
I updated the PR.

**Pull request merged**